### PR TITLE
Simplify recovery code

### PR DESCRIPTION
**Pull request by graetzer**
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr-linux/5424/

### CODE REVIEW

File: arangod/RocksDBEngine/RocksDBBuilderIndex.cpp [hunk 1]
```diff
@@ -502,10 +503,10 @@ Result catchup(RocksDBIndex& ridx, WriteBatchType& wb, AccessMode::Type mode,
   if (res.ok()) {
     numScanned = replay.numInserted + replay.numRemoved;
     res = trx.commit();  // important for iresearch
-    
-    if (ridx.estimator() != nullptr) {
-      ridx.estimator()->setCommitSeq(rootDB->GetLatestSequenceNumber());
-    }
+//
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBBuilderIndex.cpp]: is that art, or can it be removed entirely?

File: arangod/RocksDBEngine/RocksDBCollection.cpp [hunk 2]
```diff
@@ -375,30 +376,31 @@ std::shared_ptr<Index> RocksDBCollection::createIndex(VPackSlice const& info,
...
-                            _logicalCollection.id());
-    rocksdb::PinnableSlice ps;
-    rocksdb::Status s = engine->db()->Get(rocksdb::ReadOptions(),
-                                          RocksDBColumnFamily::definitions(),
-                                          key.string(), &ps);
-    if (!s.ok()) {
-      res.reset(rocksutils::convertStatus(s));
-    } else {
+  
+  do {
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBCollection.cpp]: As far as I can see, the `do...while(false)` construct here is only used to `break` out of it prematurely.
As the block is relatively long, wouldn't this justify moving it to a function/method of its own? That would allow `return`ing prematurely and doesn't require a `do..while(false)` workaround.

File: arangod/RocksDBEngine/RocksDBCollectionMeta.cpp [hunk 3]
```diff
@@ -272,16 +269,16 @@ Result RocksDBCollectionMeta::serializeMeta(rocksdb::WriteBatch& batch,
           << "beginning estimate serialization for index '" << idx->objectId() << "'";
       output.clear();
 
-      seq = est->serialize(output, maxCommitSeq);
-      if (seq > 0) {
-        // calculate retention sequence number
-        appliedSeq = std::min(appliedSeq, seq);
-      }
+      est->serialize(output, maxCommitSeq);
+//      if (seq > 0) {
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBCollectionMeta.cpp]: again, is this still needed, or can it be removed?

File: arangod/RocksDBEngine/RocksDBCuckooIndexEstimator.h [hunk 4]
```diff
@@ -153,7 +153,8 @@ class RocksDBCuckooIndexEstimator {
 
  public:
   static bool isFormatSupported(arangodb::velocypack::StringRef serialized) {
-    switch (serialized.front()) {
+    TRI_ASSERT(serialized.size() > sizeof(_appliedSeq) + sizeof(char));
+    switch (serialized[sizeof(_appliedSeq)]) {
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBCuckooIndexEstimator.h]: mmmh. Previously we switched upon the value of a single character at the first byte of the StringRef. Now we are switching upon the value of a different byte... Is that downwards-compatible, and what is the motivation for the change?

File: arangod/RocksDBEngine/RocksDBCuckooIndexEstimator.h [hunk 5]
```diff
@@ -205,9 +205,9 @@ class RocksDBCuckooIndexEstimator {
         _nrUsed(0),
         _nrCuckood(0),
         _nrTotal(0),
-        _committedSeq(commitSeq),
+        _appliedSeq(0),
         _needToPersist(false) {
-    switch (serialized.front()) {
+    switch (serialized[sizeof(_appliedSeq)]) {
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBCuckooIndexEstimator.h]: same here

File: arangod/RocksDBEngine/RocksDBCuckooIndexEstimator.h [hunk 6]
```diff
@@ -465,8 +462,8 @@ class RocksDBCuckooIndexEstimator {
   uint64_t nrCuckood() const { return _nrCuckood; }
 
   bool needToPersist() const {
-    READ_LOCKER(locker, _lock);
-    return _needToPersist.load();
+//    READ_LOCKER(locker, _lock);
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBCuckooIndexEstimator.h]: let's remove the READ_LOCKER here then
```suggestion
```

File: arangod/RocksDBEngine/RocksDBCuckooIndexEstimator.h (hunk 4 above)
Comment (graetzer) [reply arangod/RocksDBEngine/RocksDBCuckooIndexEstimator.h]: I did not design the format, my intention was to put the deserialization code into the cuckoo estimator

File: arangod/RocksDBEngine/RocksDBCuckooIndexEstimator.h (hunk 5 above)
Comment (danielhlarkin) [reply arangod/RocksDBEngine/RocksDBCuckooIndexEstimator.h]: This should be fine as-is.

File: arangod/RocksDBEngine/RocksDBCuckooIndexEstimator.h (hunk 4 above)
Comment (danielhlarkin) [reply arangod/RocksDBEngine/RocksDBCuckooIndexEstimator.h]: This should be fine as-is. Just moving the check from one place to another, format hasn't changed.

File: arangod/RocksDBEngine/RocksDBCollection.cpp (hunk 2 above)
Comment (danielhlarkin) [reply arangod/RocksDBEngine/RocksDBCollection.cpp]: I agree. The `do...while(false)` construct feels pretty hacky and anti-pattern here. I'd say either break it out into a function and use returns, or use a status variable and normal conditional logic.

File: arangod/RocksDBEngine/RocksDBCollectionMeta.cpp [hunk 7]
```diff
@@ -140,7 +140,7 @@ rocksdb::SequenceNumber RocksDBCollectionMeta::applyAdjustments(rocksdb::Sequenc
 
   decltype(_bufferedAdjs) swapper;
   {
-    std::lock_guard<std::mutex> guard(_countLock);
+    std::lock_guard<std::mutex> guard(_bufferLock);
```
[the remark this answers is not in the document]
Comment (graetzer) [reply arangod/RocksDBEngine/RocksDBCollectionMeta.cpp]: no I think its exclusive

File: arangod/RocksDBEngine/RocksDBCollectionMeta.cpp [hunk 8]
```diff
@@ -202,6 +202,7 @@ Result RocksDBCollectionMeta::serializeMeta(rocksdb::WriteBatch& batch,
   } else {  // maxCommitSeq is == UINT64_MAX without any blockers
     appliedSeq = std::min(appliedSeq, maxCommitSeq);
   }
+  TRI_ASSERT(commitSeq <= appliedSeq);
```
Comment (danielhlarkin) [inline arangod/RocksDBEngine/RocksDBCollectionMeta.cpp]: I think what you actually want is `TRI_ASSERT(appliedSeq <= commitSeq)` inside the `if` branch of the statement before the assignment.

Comment (graetzer) [reply arangod/RocksDBEngine/RocksDBCollectionMeta.cpp]: hm shouldn't it hold regardless ?
